### PR TITLE
Pipeline Prefill

### LINKED ISSUE #45836: Adjust prefill model to support D2D sockets
When we have #45829 integrate in into model

### LINKED ISSUE #45837: Per layer build transformer prefill model for multihost

### PR DESCRIPTION
**Pull request by jjovicicTT**
# Pipeline-parallel prefill for DeepSeek-V3 / Kimi-K2.6 over D2D sockets

## Summary

Adds pipeline-parallel (PP) chunked prefill for deepseek_v3_d_p, splitting the transformer across N galaxies (ranks) connected by device-to-device (D2D) socket transport over fabric. Builds on the D2D stream service: each rank owns a contiguous layer range, runs its slice of every chunk, and forwards activations to the next rank over a D2D socket; the final rank holds the populated KV cache. Runs 1, 2, and 4 ranks from the same runner and config set.

## What's included

- Pipeline prefill runner (prefill_runner.py, reworked): rank-aware layer split, two serving modes sharing one chunk loop —
  - standalone (bounded, self-contained): trace input on rank 0, D2D between ranks, per-rank KV PCC check against the golden trace.
  - request (unbounded): rank 0 reads chunks from the H2D socket driven by an external producer; runs to SIGTERM.
- Rank-sliced model: tt_prefill_transformer.py and tt_prefill_runtime.py (renamed from chunked_prefill_pipeline.py) thread first_layer_idx / is_first_rank / is_last_rank; last rank's last layer runs KV-only (no norm/LM-head tail) since chunked prefill never samples.
- Topology configs (topology_configuration/): connected mesh-graph descriptors for 2- and 4-galaxy chains, rank bindings (1/2/4-rank D2D and request variants), and a 1-galaxy/1-rank config.
- Tooling: run_pipeline_prefill.sh launcher (FABRIC_2D + connected MGD) and plot_pipeline_trace.py for per-rank chunk timeline visualization.
- Model-variant support: DeepSeek-V3 and Kimi-K2.6 (incl. Kimi's l1-small routing-semaphore path) selectable via PREFILL_MODEL_VARIANT.

## Pipeline trace visualization

plot_pipeline_trace.py renders a run's behaviour as a Gantt chart straight from its log — no extra instrumentation. x = wall-clock (s, relative to first chunk start), y = rank; each chunk is a bar on its rank's row, coloured by chunk index, so the same chunk is the same colour on every rank and you watch it travel down the ranks (the pipeline diagonal). Handoff arrows trace chunk c from rank r → r+1. It parses the per-chunk CHUNK_START epochs and the post-loop E2E_CLOCK lines from the combined MPI-tagged log, so it works on any standalone or request run.

python -m models.demos.deepseek_v3_d_p.tt.runners.plot_pipeline_trace <run.log> -o pipeline_trace.png

image

## Stress tesing

image

## Validation

- Single-galaxy standalone: per-rank KV PCC passes against the golden trace.
- 2-galaxy pipeline (half model per rank): KV PCC matches the single-rank reference; D2D validated over FABRIC_2D with a connected MGD.
- Functional across 1/2/4-rank configs.

### Standalone (benchmark / PCC — reads golden trace, no producer)

./models/demos/deepseek_v3_d_p/tt/runners/run_pipeline_prefill.sh \
  models/demos/deepseek_v3_d_p/tt/runners/topology_configuration/pipeline_prefill_rank_binding_4rank_d2d.yaml \
  "$RING" ens5f0np0
(That binding carries PREFILL_STANDALONE=1, PREFILL_MODEL_VARIANT=kimi_k2_6, PREFILL_PP_D2D_FIFO_BYTES=4096, 11 chunks — Kimi-ready as-is.)

### Request (production loop — needs an H2D producer)

Terminal A (runner, on c09u02):

./models/demos/deepseek_v3_d_p/tt/runners/run_pipeline_prefill.sh \
  models/demos/deepseek_v3_d_p/tt/runners/topology_configuration/pipeline_prefill_request_4rank.yaml \
  "$RING" ens5f0np0

Terminal B (producer on c09u02, once the runner logs [h2d] descriptor ... -> /dev/shm/...):

PREFILL_H2D_SERVICE_ID=ds_prefill PREFILL_STANDALONE_NCHUNKS=11 PREFILL_STANDALONE_ITERS=3 \
  python3 -m models.demos.deepseek_v3_d_p.tt.runners.prefill_h2d_producer

-- multi-user stress instead:  python3 -m models.demos.deepseek_v3_d_p.tt.runners.prefill_h2d_stress_producer

## Notes / follow-ups

- Migration KV-chunk-table publish + per-layer LayerAck are wired for single-rank only; pipelined migration is future work.

## CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

https://github.com/tenstorrent/tt-metal/actions/runs/28170015625

https://github.com/tenstorrent/tt-metal/actions/runs/28170025561

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:jjovicic/pipeline-prefill-d2d)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:jjovicic/pipeline-prefill-d2d)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:jjovicic/pipeline-prefill-d2d)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:jjovicic/pipeline-prefill-d2d)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:jjovicic/pipeline-prefill-d2d)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:jjovicic/pipeline-prefill-d2d)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:jjovicic/pipeline-prefill-d2d)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:jjovicic/pipeline-prefill-d2d)

### CODE REVIEW

**Review by pavlepopovic — APPROVED** [review]

**Review by nbabinTT — APPROVED** [review]

File: models/demos/deepseek_v3_d_p/scripts/plot_pipeline_trace.py
Comment (philei-tt) [inline models/demos/deepseek_v3_d_p/scripts/plot_pipeline_trace.py:1]: (nit) maybe move in `scripts/` subdirectory? Not a fan of mixing visualization tools and business logic

File: models/demos/deepseek_v3_d_p/tt/runners/prefill_kv_validation.py [hunk 1]
```diff
@@ -19,7 +19,7 @@
 
 import ttnn
 from models.demos.deepseek_v3_d_p.tt.runners.runner_utils import get_variant
-from models.demos.deepseek_v3_d_p.tt.tt_deepseek_prefill_pipeline import TtDeepSeekPrefillPipeline
+from models.demos.deepseek_v3_d_p.tt.tt_prefill_runtime import TtPrefillRuntime as TtDeepSeekPrefillPipeline
```
Comment (philei-tt) [inline models/demos/deepseek_v3_d_p/tt/runners/prefill_kv_validation.py]: Why? :D Just rename usages to get rid of confusing naming

File: models/demos/deepseek_v3_d_p/tt/runners/prefill_runner.py [hunk 2]
```diff
@@ -1,55 +1,67 @@
...
+the same TtPrefillRuntime (first_layer_idx / is_first_rank / is_last_rank). With >1 rank the cross-rank
+hidden state moves device-to-device over fabric sockets (connected MGD + FABRIC_2D); N=1 is the
+single-galaxy case (no transport). Ranks run decoupled (no per-chunk barrier; one warm-up barrier
+after compile). The two modes run identical pipeline mechanics and differ only in the trigger:
+
+  * Request mode (default): production serving. rank 0's tokens + per-iter PrefillMetadata arrive over
+    the H2D socket from an external producer (prefill_h2d_producer.py / the scheduler); the loop is
+    UNBOUNDED (runs to SIGTERM). KV-chunk-table migration + per-layer LayerAck are wired for the
+    single-rank case only (disabled for the pipeline for now). Shutdown for >1 rank is rough: ranks
+    block in the H2D/D2D recv device op and exit on teardown/SIGKILL (no end-of-request sentinel yet).
```
Comment (philei-tt) [inline models/demos/deepseek_v3_d_p/tt/runners/prefill_runner.py:17]: d2d sockets also support metadata, maybe make them propagate sigterm signal through d2d to gracefully exit? But I guess ideally it should not be a busy loop, right?

File: models/demos/deepseek_v3_d_p/tt/runners/prefill_runner.py [hunk 3]
```diff
@@ -85,197 +98,319 @@ def _handle_sigterm(signum, frame):
...
+
+def _d2d_send(outbound, activation: ttnn.Tensor, rank: int, meta: dict) -> None:
+    """Push this rank's output hidden state + metadata to the downstream rank's receiver. Coerce the
+    activation to the sender backing's memory layout first so the op's page-for-page copy lines up,
+    then free it."""
+    t0 = time.perf_counter()
+    backing = outbound.get_backing_tensor()
+    if activation.memory_config() != backing.memory_config() or activation.layout != backing.layout:
+        activation = ttnn.to_layout(activation, backing.layout)
+        activation = ttnn.to_memory_config(activation, backing.memory_config())
```
Comment (philei-tt) [inline models/demos/deepseek_v3_d_p/tt/runners/prefill_runner.py]: Maybe warning? This impacts perf, right?

File: models/demos/deepseek_v3_d_p/tt/runners/prefill_runner.py [hunk 4]
```diff
@@ -85,197 +98,319 @@ def _handle_sigterm(signum, frame):
...
+        torch.tensor(words, dtype=torch.int32).reshape(1, 1, 1, -1),
+        dtype=ttnn.uint32,
+        layout=ttnn.ROW_MAJOR_LAYOUT,
+        device=backing.device(),
+        memory_config=ttnn.DRAM_MEMORY_CONFIG,
+        mesh_mapper=ttnn.create_mesh_mapper(
+            backing.device(),
+            ttnn.MeshMapperConfig(placements=[ttnn.PlacementReplicate(), ttnn.PlacementReplicate()]),
+        ),
+    )
```
Comment (philei-tt) [inline models/demos/deepseek_v3_d_p/tt/runners/prefill_runner.py:262]: forced device sync on metadata is not great, is it worth moving this into kernel in the future? 🤔

File: models/demos/deepseek_v3_d_p/tt/runners/prefill_runner.py [hunk 5]
```diff
@@ -85,197 +98,319 @@ def _handle_sigterm(signum, frame):
...
+    """Run one chunk: prefill, forward the output downstream (non-last rank) and grant the outbound
+    sender so it ships over fabric, log CHUNK_START. Returns the compute-start epoch (NTP-comparable)."""
+    t_start = time.time()
+    out = runtime.prefill(
+        inp, slot_id=meta["slot_id"], actual_start=meta["actual_start"], actual_end=meta["actual_end"]
+    )
+    if not runtime.config.is_last_rank:
+        _d2d_send(d2d_out, out, rank, meta)  # push + free; the grant below forwards it over fabric
+    if d2d_out is not None:
+        d2d_out.release_fabric_links()
```
Comment (philei-tt) [inline models/demos/deepseek_v3_d_p/tt/runners/prefill_runner.py:293]: This looks fragile (I know it's outside the scope of this PR). @nmauriceTT I wonder why this can't be handled inside d2d socket itself? It know when it finished sending, so it can release fabric links on it's own?

**Review by philei-tt — APPROVED** [review]
Overall LGTM, thank you. Left a few comments

File: models/demos/deepseek_v3_d_p/tt/runners/prefill_runner.py (hunk 2 above)
Comment (jjovicicTT) [reply models/demos/deepseek_v3_d_p/tt/runners/prefill_runner.py:17]: yes, this needs to be fixed and made right, ideally I guess the producer (prefill scheduler in production) should send the termination event and it can be propagated to all ranks 🤷 

we will think about how to make this

File: models/demos/deepseek_v3_d_p/tt/runners/prefill_runner.py (hunk 4 above)
Comment (jjovicicTT) [reply models/demos/deepseek_v3_d_p/tt/runners/prefill_runner.py:262]: yes, there is work on this, ops that use the metadata need to be able to read from device tensors for this to be fully removed, here is the issue tracking this

https://github.com/tenstorrent/tt-metal/issues/45414

File: models/demos/deepseek_v3_d_p/tt/runners/prefill_runner.py (hunk 3 above)
Comment (jjovicicTT) [reply models/demos/deepseek_v3_d_p/tt/runners/prefill_runner.py]: checking, thanks!

Comment (jjovicicTT) [reply models/demos/deepseek_v3_d_p/tt/runners/prefill_runner.py]: confirmed this never triggers so removed this, thanks!

File: models/demos/deepseek_v3_d_p/tt/runners/prefill_runner.py (hunk 5 above)
Comment (nmauriceTT) [reply models/demos/deepseek_v3_d_p/tt/runners/prefill_runner.py:293]: _Right now_:
The D2D service knows when it is done running. But it does not know if anybody else is running. 
CCLs do not know either if a D2D service is running. 

To prevent data-races / hangs, you need a scheduler. Right now, as far as I know, it does not seem that ethernet cores handles this or provide a way to know if someone is already using it. 
That's why this synchronization is done on host, by manually granting/waiting for fabric sockets to service cores. 

I guess one solution would be to have CCLs call `release_fabric_links`/`wait_fabric_links`. But that would mean that CCLs should be aware of _all_ possible D2D services on a given device.  

( @aagarwalTT for visibility )

Comment (aagarwalTT) [reply models/demos/deepseek_v3_d_p/tt/runners/prefill_runner.py:293]: This release call here is to let the d2d service resume its operation - connect to fabric and push out the data. 

And yes, the d2d socket is aware when its done with an iteration/transfer and it releases links on its own, which is handled with the corresponding wait call.

I understand if the naming is confusing, we can look into making it more intuitive. In a nutshell - the wait call is for the model graph to wait for d2d service to relinquish the fabric links and the release call is for the d2d service to resume its work.
